{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ACS XML Parser\n",
    "This parser is used to extract the text, sequences, and meta data from ACS articles. The parsing process has two parts: First it works on the article xml files to extract the text and meta data from them, then it looks at the supplementary files for genetic sequences. At the end, the parser outputs the extracted data as pickle files for downstream tasks. For a more detailed specification on the input/output formats, please visit the [Github repo](https://github.com/synbioks/ACS-XML-to-text/tree/acs-parser-jiawei)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Environment Configuration\n",
    "To run this script you need to install the following package: `lxml` for text, metadata extraction, `pprint` and `tqdm` for debug outputs. Also you need to install `pdftotext` for pdf sequence extraction (We are using the pdftotext cli instead of importing it as a python package because cli has the option we need)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import requests\n",
    "import pickle\n",
    "import subprocess as sp\n",
    "from pprint import pprint\n",
    "from lxml import etree\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input Data\n",
    "The input to this script is the data dump from ACS. The data dump consist of folders where each folder contains one ACS article in xml format and one folder called suppl. The suppl folder has all the supplementary files related to that article.\n",
    "\n",
    "example:\n",
    "```\n",
    "sb000001           <- this is one folder, and we refer \"sb000001\" as the publication number\n",
    "|-- sb000001.xml   <- this is the article file\n",
    "+-- suppl          <- this folder contains all the supplementary file of this article\n",
    "    |-- file_1.pdf\n",
    "    +-- file_2.pdf\n",
    "```\n",
    "\n",
    "Currently there are 1597 folders like the example in the ACS data dump, and you can find those folders in `/sbksvol/data/acs-data/article-files`. We call this path the input path of the parser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script Arguments\n",
    "The parser takes three arguments:\\\n",
    "`input_path` tells the parser where to look for ACS data dump.\\\n",
    "`output_path` tells the parser where to store the extracted text, metadata, and genetic sequences of articles.\\\n",
    "`txt_path` tells the parser where to store the text of articles. \n",
    "\n",
    "The differences between `output_path` and `txt_path` are: `output_path` contains pickle files that have all the information extracted from articles and supplementary files (text, metadata, genetic sequences), whereas `txt_path` contains txt files that only have the text of articles. We did this because storing the text of articles in txt format is easier to read for tasks that do not require metadata and sequence information. The text data in both paths are identical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# arguments\n",
    "# the directory that stores our supplementary data dump\n",
    "# since the supplementary data dump also contains the original article\n",
    "# we can use this as the input path\n",
    "input_path = os.path.abspath(\"/mnt/data1/jiawei/acs-data/suppl-files/\")\n",
    "\n",
    "# the directory that stores our processed pickle files\n",
    "output_path = os.path.abspath(\"/mnt/data1/jiawei/acs-data/processed-files/\")\n",
    "\n",
    "# we also provide extracted text in plain text format, this is the output path\n",
    "# of the plain text\n",
    "txt_path = os.path.abspath(\"/mnt/data1/jiawei/acs-data/txt-files\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional Parameters\n",
    "The parser also has some additional parameters:\\\n",
    "`valid_cache_path` points to a pickle files that stores the result of sbol api queries. This is to save time (api query is very slow over the network) for repeated run/debug.\\\n",
    "`non_acs_article_path` points to a txt files that has a list of publication numbers. Articles having these publication numbers are non-research article. \n",
    "\n",
    "**Important**: `non_acs_article_path` is deprecated because the list is not accurate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# local file parameters don't change unless you know what you are doing\n",
    "\n",
    "# sbol api query cache, use to speed up the process time\n",
    "valid_cache_path = os.path.abspath(\"./sbol-validation-cache.pkl\")\n",
    "\n",
    "# deprecated, the list is inaccurate (false negative, non research article not in the list)\n",
    "non_acs_article_path = os.path.abspath(\"./non-acs-article.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SBOL API\n",
    "The [sbol api](http://synbiodex.github.io/SBOL-Validator/#introduction) is an online service provided by [Myers Research Group at University of Utah](https://async.ece.utah.edu/tools/sbol-validatorconverter/). It takes any file, checks if the file is a *valid sequence file* (meaning: a file contains sequences information and adheres to one of the many popular computer-readable formats), and converts the sequence file into one of the following four formats (fasta, genbank, sbol, sbol2). We utilize this api in two ways: We use it to check whether a supplementary file is a valid sequence file, and convert a valid sequence file in any format to [fasta format](https://en.wikipedia.org/wiki/FASTA_format).\n",
    "\n",
    "Through an inspection at the supplementary files, we found that files that have extension like \".gb\", \".sbol\", \".fasta\", and \".dna\" are valid sequence files, \".txt\" supplementary files can be genbank files, and \".xml\" supplementary files can be sbol files. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sbol api request params\n",
    "sbol_validator_url = \"https://validator.sbolstandard.org/validate/\"\n",
    "\n",
    "# these are the file format \n",
    "sbol_allowed_file_type = set([\n",
    "    \"gb\", \"fasta\", \"sbol\", \"txt\", \"xml\", \"dna\"\n",
    "])\n",
    "validator_param = {\n",
    "    'options': {\n",
    "        'language' : \"FASTA\",\n",
    "        'test_equality': False,\n",
    "        'check_uri_compliance': False,\n",
    "        'check_completeness': False,\n",
    "        'check_best_practices': False,\n",
    "        'fail_on_first_error': True,\n",
    "        'provide_detailed_stack_trace': False,\n",
    "        'subset_uri': '',\n",
    "        'uri_prefix': 'dummy', # we need to have something here or the api will return an error\n",
    "        'version': '',\n",
    "        'insert_type': False\n",
    "    },\n",
    "    \"main_file\": None,\n",
    "    \"return_file\": True\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Text and Metadata Extraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1.1: Helper Functions for Extraction\n",
    "\n",
    "**Note**: To better understand the code below, I strongly suggest you to first read [lxml documentation](https://lxml.de/tutorial.html) and [XPath 1.0 documentation](https://www.w3.org/TR/1999/REC-xpath-19991116/), and open an article xml file on the side when going through this section (I used sb9b00030.xml as a reference when writing this documentation).\n",
    "\n",
    "The following code cells are many helper functions help the text and metadata extraction process. Currently, the parser have the following features:\n",
    "\n",
    "- **Extract Text:** The text of the articles are wrapped in `<p>` and `<title>` tags inside the xml files following the xpath `/article/body`. We use `extract_body()` to find those tags. An example is shown below:\n",
    "```\n",
    "<p>\n",
    "    Spontaneous and random back and forth switching between states \n",
    "    is certainly not suitable for processes requiring irreversible\n",
    "    cell fate determination, such as development and cell \n",
    "    differentiation. In these contexts, the intracellular noise \n",
    "    would need to be low enough to avoid spontaneous state \n",
    "    switching. To identify possible ratio control strategies in \n",
    "    such a low-noise environment, we transitioned experimentally \n",
    "    to a less-noisy system: the chromosomally integrated mutual \n",
    "    inhibition toggle in\n",
    "    <italic toggle=\"yes\">S. cerevisiae</italic>\n",
    "    .\n",
    "    <xref rid=\"ref25\" ref-type=\"bibr\"/>\n",
    "    This circuit has the same topology as that shown in\n",
    "    <xref rid=\"fig1\" ref-type=\"fig\">\n",
    "    Figure\n",
    "    <xref rid=\"fig1\" ref-type=\"fig\"/>\n",
    "    </xref>\n",
    "    a, exhibits hysteretic behavior (\n",
    "    <xref rid=\"fig2\" ref-type=\"fig\">\n",
    "    Figure\n",
    "    <xref rid=\"fig2\" ref-type=\"fig\"/>\n",
    "    </xref>\n",
    "    a), and favors the TetR-dominant, low-GFP state under no induction. \n",
    "    However, differences in promoters, copy number, and transcription\n",
    "    –translation processes between\n",
    "    <italic toggle=\"yes\">E. coli</italic>\n",
    "    and yeast serve to reduce intracellular noise and shift the \n",
    "    bistable region up to roughly 3–17 ng/mL ATc. Unlike in\n",
    "    <italic toggle=\"yes\">E. coli</italic>\n",
    "    , the bulk of the bistable range was impervious to the effects \n",
    "    of intracellular noise, resulting in a single peak homogeneous \n",
    "    expression profile even when the system is operating within the \n",
    "    bistable region.\n",
    "</p>\n",
    "```\n",
    "In general the text contains formatting tags, embedded references, and captions. Our goal is to remove these and extract the plain text. To do this, we implement `p_helper()` to retrieve the plain text in fragments, join them together into one line, and fix punctuation/parenthesis placement. We call this one line of sanitized text returned by `p_helper()` a *paragraph*. `title_helper()` is used to retrive the title in plain text. In addition to extracting paragraphs and titles, we also use the titles to label what section (intro, method, results, discussion, and so on) the paragraphs are in. The logic is coded in `extract_body()`: If we see a paragraph following a title containing \"method\", we indicate the paragraph is in \"method\" section.\n",
    "\n",
    "\n",
    "- **Extract Abstract:** Abstract is found in tag `<abstract>`. It is wrapped in a `<p>` tag similar to a paragraph in the body. We use `extract_abstract()` to locate the tag and `p_helper()` to extract the plain text.\n",
    "\n",
    "\n",
    "- **Extract Keywords:** Keywords can be found in tag `<kwd-group>`. `extract_keywords()` and `kwd_helper()` is used to locate and retrieve the keywords in the article.\n",
    "\n",
    "\n",
    "- **Extract Publication Dates:** There are two type of dates we can find in the metadata, one is electronic publication date and the other is issue puiblication date, they are stored in `<pub-date>` tags. You can tell which is the electronic publication date by looking at the attributes of the tags. `extract_date()` handles dates extraction.\n",
    "\n",
    "\n",
    "- **Extract History of Events:** Stored in `<history>`, history of events is a list of event the article has been through. Each entry in the list contains the name of event and the date of the event. `extract_history()` handles history extraction.\n",
    "\n",
    "\n",
    "- **Extract Article Type:** Located in `<subj-group>`, It indicate the type of the article, and we are currently using this to determine if an article is research article or not. `extract_article_type()` handles this process.\n",
    "\n",
    "\n",
    "- **Extract id and internal id:** Overall, we can find three different ids that can uniquely identify an article: 1) the publication number (mentioned in \"Input Data\" section above), 2) the id in `<article-id>`, and 3) the internal id in the attribute of `<article>`. We believe 3) is an id used internally by ACS because we cannot find any public available information using this id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to collect matching files and dirs\n",
    "# basically a fancy find method\n",
    "# pattern is a re pattern\n",
    "# collect_dirs is a bool, when set to true it will collect directories/folders\n",
    "# return a list of files/directories that matches the pattern\n",
    "def collect_files(root, res, pattern=\"\", collect_dirs=True, min_depth=None, max_depth=None):\n",
    "    \n",
    "    # check max depth\n",
    "    if not max_depth is None and max_depth == 0:\n",
    "        return\n",
    "    \n",
    "    # go through all item in the dir\n",
    "    for item in os.listdir(root):\n",
    "        \n",
    "        # process item\n",
    "        item_path = os.path.join(root, item)\n",
    "        item_is_dir = os.path.isdir(item_path)\n",
    "        \n",
    "        # put valid file in res if min depth has reached\n",
    "        if min_depth is None or min_depth - 1 <= 0:\n",
    "            if re.match(pattern, item_path):\n",
    "                if not item_is_dir or collect_dirs:\n",
    "                    res.append(item_path)\n",
    "        \n",
    "        # recursively collect all files\n",
    "        if item_is_dir:\n",
    "            next_min_depth = None if min_depth is None else min_depth - 1\n",
    "            next_max_depth = None if max_depth is None else max_depth - 1\n",
    "            collect_files(item_path, res, pattern, collect_dirs, next_min_depth, next_max_depth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helps to extract text from paragraph\n",
    "def p_helper(node):\n",
    "    \n",
    "    # <p/> does not have text\n",
    "    if node.text is None:\n",
    "        return \"\"\n",
    "    \n",
    "    # each paragarph is put into a line\n",
    "    line_list = [node.text]\n",
    "    for child in node:\n",
    "\n",
    "        # get the text inside the child if the tag isn't \n",
    "        # named-content and inline-formula (those are mathametical formulas)\n",
    "        # and the text following the child\n",
    "        if not child.tag in (\"named-content\", \"inline-formula\"):\n",
    "            line_list.append(\" \".join(child.xpath(\".//text()\")))\n",
    "        line_list.append(child.tail)\n",
    "\n",
    "    # there might be none in line_list\n",
    "        \n",
    "    # re dark magic below\n",
    "    # remove new line and spaces\n",
    "    line = \" \".join(line_list)\n",
    "    line = line.strip()\n",
    "    line = line.replace(\"\\n\", \" \")\n",
    "\n",
    "    # clean up consecutive spaces\n",
    "    line = re.sub(\"\\s+\", \" \", line)\n",
    "\n",
    "    # fix the space around punctuation\n",
    "    line = re.sub(\"\\s([.,\\):;])\", r\"\\1\", line)\n",
    "    line = re.sub(\"\\(\\s\", r\"(\", line)\n",
    "    line = re.sub(\"\\s*([-/])\\s*\", r\"\\1\", line)\n",
    "    return line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# strip format from keyword nodes\n",
    "def kwd_helper(node):\n",
    "    \n",
    "    # return a keyword string\n",
    "    kwd_tokens = node.xpath(\".//text()\")\n",
    "    kwd = \" \".join(kwd_tokens).replace(\"\\n\", \" \").strip()\n",
    "    kwd = re.sub(\"\\s+\", \" \", kwd)\n",
    "    return kwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this returns interesting titles\n",
    "# for example: intro, method, and results\n",
    "# return None for non interesting titles\n",
    "def title_helper(node):\n",
    "    \n",
    "    # extract text from title node\n",
    "    title = \" \".join(node.xpath(\".//text()\"))\n",
    "    title = title.replace(\"\\n\", \" \")\n",
    "    title = re.sub(\"\\s+\", \" \", title)\n",
    "    title = title.strip()\n",
    "    title = title.lower()\n",
    "    \n",
    "    # categorize title\n",
    "    res = []\n",
    "    if \"intro\" in title:\n",
    "        res.append(\"introduction\")\n",
    "    if \"result\" in title:\n",
    "        res.append(\"result\")\n",
    "    if \"discuss\" in title:\n",
    "        res.append(\"discussion\")\n",
    "    if \"material\" in title:\n",
    "        res.append(\"materials\")\n",
    "    if \"method\" in title or \"procedure\" in title:\n",
    "        res.append(\"method\")\n",
    "    if \"summary\" in title:\n",
    "        res.append(\"summary\")\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract text from xpath nodes\n",
    "def extract_body(root):\n",
    "    \n",
    "    # we are interested in the text in the body section\n",
    "    curr_title = []\n",
    "    text = []\n",
    "    text_nodes = root.xpath(\"/article/body//*[self::p or (self::title and not(ancestor::caption))]\")\n",
    "    for text_node in text_nodes:\n",
    "        \n",
    "        # handle title\n",
    "        if text_node.tag == \"title\":\n",
    "            tmp_title = title_helper(text_node)\n",
    "            if len(tmp_title) > 0:\n",
    "                curr_title = tmp_title\n",
    "            title = \" \".join(text_node.xpath(\".//text()\"))\n",
    "            title = title.replace(\"\\n\", \" \")\n",
    "            title = re.sub(\"\\s+\", \" \", title)\n",
    "            title = title.strip()\n",
    "            text.append({\n",
    "                \"text\": title,\n",
    "                \"section\": curr_title\n",
    "            })\n",
    "        \n",
    "        # handle paragraph\n",
    "        elif text_node.tag == \"p\":\n",
    "            text.append({\n",
    "                \"text\": p_helper(text_node),\n",
    "                \"section\": curr_title\n",
    "            })\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract abstract\n",
    "def extract_abstract(root):\n",
    "    \n",
    "    # get the abstract paragraph\n",
    "    abstract = []\n",
    "    abstract_nodes = root.xpath(\"//abstract/p\")\n",
    "    if abstract_nodes:\n",
    "        abstract.append(p_helper(abstract_nodes[0]))\n",
    "    return abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract_keywords from meta data\n",
    "def extract_keywords(root):\n",
    "    \n",
    "    # get the keywords\n",
    "    keywords = []\n",
    "    kwd_nodes = root.xpath(\"//kwd-group/kwd\")\n",
    "    for kwd_node in kwd_nodes:\n",
    "        keywords.append(kwd_helper(kwd_node))\n",
    "    return keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract date information from meta data\n",
    "def extract_date(root):\n",
    "    \n",
    "    issue_pub_date = None\n",
    "    electron_pub_date = None\n",
    "    \n",
    "    # traverse to the date note\n",
    "    date_nodes = root.xpath(\"/article/front/article-meta/pub-date\")\n",
    "    \n",
    "    # get the time\n",
    "    for node in date_nodes:\n",
    "        year = node.xpath(\"./year\")[0].text.strip()\n",
    "        month = node.xpath(\"./month\")[0].text.strip()\n",
    "        day = node.xpath(\"./day\")[0].text.strip()\n",
    "\n",
    "        if \"date-type\" in node.attrib and node.attrib[\"date-type\"] == \"issue-pub\":\n",
    "            issue_pub_date = \"%s/%s/%s\" % (month, day, year)\n",
    "        else:\n",
    "            electron_pub_date = \"%s/%s/%s\" % (month, day, year)\n",
    "    \n",
    "    return issue_pub_date, electron_pub_date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract article-id from meta data\n",
    "def extract_id(root):\n",
    "    \n",
    "    id_node = root.xpath(\"/article/front/article-meta/article-id\")[0]\n",
    "    article_id = id_node.text.strip()\n",
    "    return article_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract internal id, the one in the first line of the xml file\n",
    "def extract_internal_id(root):\n",
    "    \n",
    "    id_node = root.xpath(\"/article\")[0]\n",
    "    return id_node.attrib[\"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract a list of history from the meta data\n",
    "def extract_history(root):\n",
    "    \n",
    "    res = []\n",
    "    dates = root.xpath(\"/article/front/article-meta/history/date\")\n",
    "    for date in dates:\n",
    "        year = date.xpath(\"./year\")[0].text.strip()\n",
    "        month = date.xpath(\"./month\")[0].text.strip()\n",
    "        day = date.xpath(\"./day\")[0].text.strip()\n",
    "        res.append({\n",
    "            \"event\": date.attrib[\"date-type\"],\n",
    "            \"time\": \"%s/%s/%s\" % (month, day, year)\n",
    "        })\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the subject of the article from meta data\n",
    "def extract_article_type(root):\n",
    "    article_type = root.xpath(\"/article/front/article-meta/article-categories/subj-group\")\n",
    "    if len(article_type) > 1:\n",
    "        print(\"article have 2 or more types\")\n",
    "    raw = article_type[0].xpath(\".//text()\")\n",
    "    # clean up\n",
    "    res = ''.join(raw).strip()\n",
    "    res = re.sub(\"\\n\", \"\", res)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1.2: Extraction Procedure\n",
    "\n",
    "1. **Collect xml Files:** `colect_files()` is used to find all the article xml files. Notice we are only collecting xml file that is one directory deep because the suppl folder also contains \n",
    "2. **Extract Information:** We open each xml file, and use the helper methods above to extract information we want.\n",
    "\n",
    "All information is stored in a dictionary call `processed_files`. It maps publication numbers to extracted information from articles. We will refer the values of this dictionary as *article info*. For example `processed_files[\"sb9b00030\"]` returns the article info retrieved from sb9b00030.xml."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total xml files: 1597\n"
     ]
    }
   ],
   "source": [
    "# collect all xml files\n",
    "xml_paths = []\n",
    "collect_files(input_path, xml_paths, pattern=\".*\\.xml$\", collect_dirs=False, min_depth=1, max_depth=2)\n",
    "print(f\"total xml files: %d\" % len(xml_paths))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deprecated: get a list of non research article\n",
    "non_acs_article_list = set()\n",
    "with open(non_acs_article_path, \"r\") as f:\n",
    "    for line in f:\n",
    "        non_acs_article_list.add(line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9ab4e0fef73d49ebb45c9467f3b750b9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1597.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# parse the files\n",
    "# after this block, the information will be extracted to var processeed_files\n",
    "# processed_files: key(pub_num) -> value(article_info)\n",
    "processed_files = {}\n",
    "for xml_path in tqdm(xml_paths):\n",
    "\n",
    "    # print(\"\\nparsing %s\" % xml_path)\n",
    "    pub_num = xml_path.split(\"/\")[-1].split(\".\")[0]\n",
    "\n",
    "    # get the root of the xml\n",
    "    root = etree.parse(xml_path).getroot()\n",
    "    \n",
    "    # get the pub date\n",
    "    issue_pub_date, electron_pub_date = extract_date(root)\n",
    "    \n",
    "    # get the article type\n",
    "    article_type = extract_article_type(root)\n",
    "\n",
    "    # create a dictionary holding the extracted data\n",
    "    xml_data = {\n",
    "        \"is_research\": \"research\" in article_type.lower(),\n",
    "        \"keywords\": extract_keywords(root),\n",
    "        \"abstract\": extract_abstract(root),\n",
    "        \"body\": extract_body(root),\n",
    "        \"issue_pub_date\": issue_pub_date,\n",
    "        \"electron_pub_date\": electron_pub_date,\n",
    "        \"article_id\": extract_id(root),\n",
    "        \"internal_id\": extract_internal_id(root),\n",
    "        \"history\": extract_history(root),\n",
    "        \"type\": article_type\n",
    "    }\n",
    "\n",
    "    # save the data\n",
    "    processed_files[pub_num] = xml_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'abstract': ['Robust and precise ratio control of heterogeneous phenotypes '\n",
      "              'within an isogenic population is an essential task, especially '\n",
      "              'in the development and differentiation of a large number of '\n",
      "              'cells such as bacteria, sensory receptors, and blood cells. '\n",
      "              'However, the mechanisms of such ratio control are poorly '\n",
      "              'understood. Here, we employ experimental and mathematical '\n",
      "              'techniques to understand the combined effects of signal '\n",
      "              'induction and gene expression stochasticity on phenotypic '\n",
      "              'multimodality. We identify two strategies to control phenotypic '\n",
      "              'ratios from an initially homogeneous population, suitable '\n",
      "              'roughly to high-noise and low-noise intracellular environments, '\n",
      "              'and we show that both can be used to generate precise '\n",
      "              'fractional differentiation. In noisy gene expression contexts, '\n",
      "              'such as those found in bacteria, induction within the circuit’s '\n",
      "              'bistable region is enough to cause noise-induced bimodality '\n",
      "              'within a feasible time frame. However, in less noisy contexts, '\n",
      "              'such as tightly controlled eukaryotic systems, spontaneous '\n",
      "              'state transitions are rare and hence bimodality needs to be '\n",
      "              'induced with a controlled pulse of induction that falls outside '\n",
      "              'the bistable region. Finally, we show that noise levels, system '\n",
      "              'response time, and ratio tuning accuracy impose trade-offs and '\n",
      "              'limitations on both ratio control strategies, which guide the '\n",
      "              'selection of strategy alternatives.'],\n",
      " 'article_id': '10.1021/acssynbio.9b00030',\n",
      " 'body': [{'section': [],\n",
      "           'text': 'Ratio control of differentiation within isogenic '\n",
      "                   'populations is a ubiquitous but poorly understood '\n",
      "                   'phenomenon. From single-celled microbes to higher '\n",
      "                   'organisms, many processes require mixed populations to '\n",
      "                   'carry out complex functions, such as bacterial '\n",
      "                   'persistence, bacterial competence, nasal and ocular '\n",
      "                   'receptor development, differentiation of blood and '\n",
      "                   'vascular cells, and immune response, as well as stem cell '\n",
      "                   'maintenance and differentiation. Several general '\n",
      "                   'explanations for how this phenotypic diversity arises have '\n",
      "                   'been proposed, such as stochastic fluctuations within gene '\n",
      "                   'regulatory networks, asymmetrical sequestering of '\n",
      "                   'regulatory proteins during cell division, and differential '\n",
      "                   'response to spatial gradients of extracellular soluble '\n",
      "                   'factors. While each of these methods could theoretically '\n",
      "                   'generate a mixture of differentiated cells within a '\n",
      "                   'population, research has frequently focused on the types '\n",
      "                   'of cells yielded, rather than the quantitative control of '\n",
      "                   'their ratios. These processes are often tightly controlled '\n",
      "                   'in terms of ratio accuracy and attainment speed to avoid '\n",
      "                   'overspecialization or to ensure normal development. '\n",
      "                   'Developing a unified understanding of the mechanisms and '\n",
      "                   'relevant factors to achieve and maintain precise ratio '\n",
      "                   'control will have widespread benefits in areas such as '\n",
      "                   'countering bacterial immunity, treating diseases in which '\n",
      "                   'ratio control is disrupted, as is the case in some '\n",
      "                   'platelet disorders or mastocytosis, or in developing '\n",
      "                   'improved protocols for stem cell differentiation.'},\n",
      "          {'section': [],\n",
      "           'text': 'As with many biological systems, the highly interconnected '\n",
      "                   'nature of the underlying genetic regulatory circuitry '\n",
      "                   'makes it difficult to study phenotypic ratio control '\n",
      "                   'without encountering myriad confounding variables. '\n",
      "                   'Synthetic biology, by cutting through to the basics of '\n",
      "                   'transcriptional regulation in isolated and orthogonal '\n",
      "                   'circuits, offers an attractive route for exploring '\n",
      "                   'mechanisms underlying ratio tuning. Studying fundamental '\n",
      "                   'genetic motifs in isolation has yielded a greater '\n",
      "                   'understanding of key cellular behaviors, such as '\n",
      "                   'multistability, oscillation, and adaptation. Multistable '\n",
      "                   'networks, specifically, are highly relevant for cellular '\n",
      "                   'differentiation processes and have wide applicability in a '\n",
      "                   'diverse range of contexts, from developmental biology to '\n",
      "                   'targeted therapeutics and cell-environment interactions. '\n",
      "                   'Synthetic toggle switches have been implemented repeatedly '\n",
      "                   'in multiple organisms, demonstrating the feasibility of '\n",
      "                   'studying differentiation with minimal, synthetic circuits. '\n",
      "                   'Further exploration of bistable circuits can reveal '\n",
      "                   'mechanisms of subpopulation control and manipulation, '\n",
      "                   'limitations of diverse control schema, and best practices '\n",
      "                   'for state transition control on single cell and population '\n",
      "                   'levels. For example, recent work has studied how circuit '\n",
      "                   'component selection affects the network’s hysteretic '\n",
      "                   'region and how inclusion of additional circuit components '\n",
      "                   'can be used to adjust population ratios.'},\n",
      "          {'section': [],\n",
      "           'text': 'Here, we present methods for reliably tuning a multimodal '\n",
      "                   'population’s ratios without the need for additional '\n",
      "                   'network components. First, using E. coli, we show that '\n",
      "                   'positioning a population with noisy intracellular '\n",
      "                   'expression dynamics within the bistable region can be '\n",
      "                   'enough to achieve fractional differentiation. Then, using '\n",
      "                   'the less noisy expression dynamics of S. cerevisiae, we '\n",
      "                   'demonstrate that temporary deviations from the bistable '\n",
      "                   'region can direct robust ratio differentiation in a '\n",
      "                   'low-noise system. By precisely modulating the stimulus '\n",
      "                   'strength and duration which moved the network toward the '\n",
      "                   'other state, we achieve reliable ratio tuning. From these '\n",
      "                   'findings, we develop a mathematical framework through '\n",
      "                   'which we can fully understand the roles of stimulus '\n",
      "                   'dosage, stimulus duration, and noise in driving fractional '\n",
      "                   'state switching of cellular populations. Gene expression '\n",
      "                   'noise acts as a global regulator of the speed and accuracy '\n",
      "                   'of ratio tuning, with noise level positively correlated '\n",
      "                   'with speed but negatively correlated with accuracy. Low '\n",
      "                   'noise systems become candidates for pulsed induction ratio '\n",
      "                   'tuning, in which an inverse correlation between stimulus '\n",
      "                   'dose and duration is observed, translating into a '\n",
      "                   'trade-off between the speed to attaining a chosen ratio '\n",
      "                   'and ratio accuracy.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Results and Discussion'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'To investigate mechanisms of ratio control in bacteria, we '\n",
      "                   'first use the well-established bistable switch circuit. '\n",
      "                   'The topology is that of mutual inhibition (Figure a) in '\n",
      "                   'which TetR and LacI repress one another while LacI is '\n",
      "                   'coexpressed with green fluorescent protein (GFP). TetR '\n",
      "                   'activity can be modulated with the addition of the small '\n",
      "                   'molecule anhydrotetracycline (ATc), which inhibits TetR '\n",
      "                   'activity and hence alleviates its inhibition on the '\n",
      "                   'expression of LacI. When grown in the absence of '\n",
      "                   'induction, the system favors the TetR dominant, low GFP '\n",
      "                   'state. A hysteresis curve was generated as a function of '\n",
      "                   'ATc concentration (Figure S1) to probe the cells’ state '\n",
      "                   'distribution in and out of its bistable region. A fitting '\n",
      "                   'with a deterministic model indicated that the left '\n",
      "                   'bifurcation point is less than 0 ng/mL ATc, while the '\n",
      "                   'right bifurcation point is approximately 2 ng/mL ATc. '\n",
      "                   'Next, we developed a stochastic model (Table S1) for the '\n",
      "                   'system which shows strong agreement with the data (Figure '\n",
      "                   'b and Figure S2). Within the bistable region bimodal '\n",
      "                   'behavior was observed, with a portion of cells in the high '\n",
      "                   'GFP state while some were in the low GFP state (Figure c). '\n",
      "                   'This bimodality is believed to be caused by noise driven '\n",
      "                   'spontaneous state transitions. Further long-term '\n",
      "                   'experiments and analyses were conducted to ensure the '\n",
      "                   'results were not affected by confounding factors (Figure '\n",
      "                   'S3).'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Gene expression noise in E. coli induces fractional '\n",
      "                   'differentiation within the bistable region. (a) Diagram of '\n",
      "                   'the mutual inhibition bistable switch. (b) In E. coli, '\n",
      "                   'this bistable switch exhibits hysteresis from 0 to roughly '\n",
      "                   '1.5 ng/mL ATc. Within this range, cells can stably hold '\n",
      "                   'either the low GFP or high GFP states. The plot shows the '\n",
      "                   'mean of three replicates’ medians ± SD, overlaid with '\n",
      "                   'model-predicted mean GFP values of 1000 simulated cells. '\n",
      "                   'Colored dots and squares are experimental results, and '\n",
      "                   'solid lines are simulation results. (c) Flow cytometry '\n",
      "                   'histograms of initially GFP-off populations show that as '\n",
      "                   'ATc dose increases, gene expression noise causes an '\n",
      "                   'increasing fraction of cells to spontaneously turn on. '\n",
      "                   'Dashed lines indicate an empirical threshold between OFF '\n",
      "                   'and ON cells. Three replicates are shown on each plot with '\n",
      "                   'corresponding dosage and the total percentage of ON cells '\n",
      "                   'noted. Data of 10 000 cells are collected for each '\n",
      "                   'experiment. (d) Quasi-potentials computed for different '\n",
      "                   'ATc dosages within the hysteretic region show the relative '\n",
      "                   'stability of the two steady states. At 0.5 ng/mL cells '\n",
      "                   'transition at similar rates between wells (arrows), but as '\n",
      "                   'ATc dosage increases to 1.5 ng/mL, the potential well for '\n",
      "                   'the low state becomes shallower, allowing cells to more '\n",
      "                   'easily transition to the high GFP state (represented by '\n",
      "                   'thicker arrowhead). Potential wells are generated from a '\n",
      "                   'stochastic simulation fit to the hysteresis curve data. '\n",
      "                   '(e) Stochastic model predicted ON percentage (blue line) '\n",
      "                   'fit experimental results (red square) accurately. The ON '\n",
      "                   'percentage is computed as the mean ON percentage of three '\n",
      "                   'replicates experimentally and that of 1000 simulated cells '\n",
      "                   'computationally for each dose of ATc.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'To quantitatively understand the relationship between '\n",
      "                   'noise and the resulting bimodality of phenotypes, the '\n",
      "                   'stochastic model was used to explore the quasipotential '\n",
      "                   'landscape underlying this bistability. The quasipotential '\n",
      "                   'landscape is estimated by the probability distribution '\n",
      "                   'from the stochastic simulation of 1000 cells. It is shown '\n",
      "                   'that as ATc concentration increases, the depth of the left '\n",
      "                   '(low GFP) potential well decreases (Figure d). This '\n",
      "                   'essentially lowers the barrier of state transition from '\n",
      "                   'low GFP to high GFP, indicating that higher ATc induction '\n",
      "                   'would result in a larger percentage of cells in the high '\n",
      "                   'GFP state. Experiments indeed confirmed model predictions '\n",
      "                   '(Figure c, Figure S4a). It is noted that the main function '\n",
      "                   'of ATc is to stimulate the production of LacI by '\n",
      "                   'suppressing the inhibition of TetR on LacI and thus change '\n",
      "                   'the depth of the two steady states’ quasipotential wells. '\n",
      "                   'Although ATc does not have a direct effect on the '\n",
      "                   'intracellular noise, it changes the amount of LacI and '\n",
      "                   'TetR, which determine the intracellular noise directly. '\n",
      "                   'This fraction changed little between measurements at 12 '\n",
      "                   'and 15 h, suggesting it is stable. Furthermore, it is '\n",
      "                   'shown that our model can predict such a ratio control with '\n",
      "                   'high quantitative accuracy (Figure e, Figure S4b,c), '\n",
      "                   'showing that high-GFP cell ratios increase monotonically '\n",
      "                   'but nonlinearly as ATc induction increases. Such a gradual '\n",
      "                   'and steady increase of ON cell percentage as a function of '\n",
      "                   'induction dosage could enable a precise ratio control for '\n",
      "                   'the entire bacterial population.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'This finding of tunable ratio control by adjusting '\n",
      "                   'induction strength within the bistable region suggests one '\n",
      "                   'possible ratio control strategy for systems that require '\n",
      "                   'precise, but uncoordinated, fractional control of '\n",
      "                   'population differentiation. Bacteria, for example, could '\n",
      "                   'keep a subset of the colony in a dormant persister state '\n",
      "                   'to ensure survival in the case of unexpected environmental '\n",
      "                   'shifts or antibiotics. Stochastic switching provides a '\n",
      "                   'simple mechanism for entering and exiting this state. A '\n",
      "                   'similar behavior is seen in the bacterial motility and '\n",
      "                   'adhesion decisions, or in switching to a mutable, '\n",
      "                   'competent state. In these cases, the overall ratio remains '\n",
      "                   'relatively fixed; though, individual cell’s states are '\n",
      "                   'not. However, this strategy demands relatively high '\n",
      "                   'intracellular noise, which is common in plasmid-based '\n",
      "                   'bacterial systems, so that individual cells could '\n",
      "                   'spontaneously transition between states within a '\n",
      "                   'reasonable amount of time.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Spontaneous and random back and forth switching between '\n",
      "                   'states is certainly not suitable for processes requiring '\n",
      "                   'irreversible cell fate determination, such as development '\n",
      "                   'and cell differentiation. In these contexts, the '\n",
      "                   'intracellular noise would need to be low enough to avoid '\n",
      "                   'spontaneous state switching. To identify possible ratio '\n",
      "                   'control strategies in such a low-noise environment, we '\n",
      "                   'transitioned experimentally to a less-noisy system: the '\n",
      "                   'chromosomally integrated mutual inhibition toggle in S. '\n",
      "                   'cerevisiae. This circuit has the same topology as that '\n",
      "                   'shown in Figure a, exhibits hysteretic behavior (Figure '\n",
      "                   'a), and favors the TetR-dominant, low-GFP state under no '\n",
      "                   'induction. However, differences in promoters, copy number, '\n",
      "                   'and transcription–translation processes between E. coli '\n",
      "                   'and yeast serve to reduce intracellular noise and shift '\n",
      "                   'the bistable region up to roughly 3–17 ng/mL ATc. Unlike '\n",
      "                   'in E. coli, the bulk of the bistable range was impervious '\n",
      "                   'to the effects of intracellular noise, resulting in a '\n",
      "                   'single peak homogeneous expression profile even when the '\n",
      "                   'system is operating within the bistable region. The noise '\n",
      "                   'level is set by the system’s size Ω in Gillespie’s '\n",
      "                   'stochastic simulation, which is widely used to control the '\n",
      "                   'noise in the gene regulator network. Though noise does not '\n",
      "                   'determine the gene expression profile during the course of '\n",
      "                   'a typical experiment, if the stochastic simulation is run '\n",
      "                   'long enough, it indeed recapitulates the data well, as '\n",
      "                   'seen in Figure S5. Instead, to induce bimodality, we '\n",
      "                   'hypothesized that internal variability could be utilized '\n",
      "                   'by temporarily forcing the cells outside of the bistable '\n",
      "                   'region favoring another state. Then the population would '\n",
      "                   'begin to transition to the other state. However, natural '\n",
      "                   'stochasticity would cause some cells to transition faster '\n",
      "                   'than others. When the population was returned to the '\n",
      "                   'bistable region prior to full-population transition, some '\n",
      "                   'fraction of the cells would finish their transition while '\n",
      "                   'the rest would return to their original state. Figure b '\n",
      "                   'schematically illustrates this process and how the various '\n",
      "                   'stimulus levels adjust the underlying potential landscape '\n",
      "                   'to induce bimodality. This diagram also shows the two '\n",
      "                   'variables which determine population response: dose, which '\n",
      "                   'measures the magnitude of the induction pulse, and '\n",
      "                   'duration, the length of time for which the dose is '\n",
      "                   'applied. Using this method, the population achieved a '\n",
      "                   'specific phenotypic ratio and individual cells only '\n",
      "                   'transition between states once, which is distinct from E. '\n",
      "                   'coli.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'In S. cerevisiae, gene expression noise is lower, '\n",
      "                   'requiring temporary deviation from the bistable region to '\n",
      "                   'induce multimodality. (a) The hysteresis curve of the '\n",
      "                   'yeast toggle switch shows strong stability. The plot shows '\n",
      "                   'mean of three flow cytometry replicates’ medians ± SD, '\n",
      "                   'overlaid with model-predicted hysteresis curve. (b) Yeast '\n",
      "                   'cells were exposed to pulses of ATc with varying dose and '\n",
      "                   'duration. Before the pulse (Phase I), all cells resided in '\n",
      "                   'the low-GFP state without any induction, corresponding to '\n",
      "                   'a single-welled potential landscape. The induction pulse '\n",
      "                   '(Phase 2) changes the underlying landscape to a single '\n",
      "                   'well in the high-GFP state, and cells begin to transition '\n",
      "                   'from low to high GFP state. Before all cells transition, '\n",
      "                   'induction is reduced to 8 ng/mL (Phase 3), at which there '\n",
      "                   'are two deep potential wells, low GFP and high GFP. '\n",
      "                   'Partially transitioned cells either transition fully or '\n",
      "                   'return to their initial low-GFP state. (c) Four flow '\n",
      "                   'cytometry experiments with four replicates each after '\n",
      "                   'pulse induction are shown with the total percentage of '\n",
      "                   'cells to the right of the dashed line. Inducer pulses for '\n",
      "                   'controlled durations allow for a wide range of ratios, '\n",
      "                   'with multiple paths to the same end point. Pulses of '\n",
      "                   'either 25 ng/mL or 30 ng/mL ATc for 4 or 8 h achieve a '\n",
      "                   'range of final ratios, with the 25 × 8 and 30 × 4 pulses '\n",
      "                   'producing nearly identical outcomes.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Systematic temporal induction experiments were then '\n",
      "                   'designed and carried out to test our hypothesis. Using '\n",
      "                   'doses of 20, 25, 30, 35, and 40 ng/mL ATc, with pulse '\n",
      "                   'durations between 2 and 24 h, we comprehensively explored '\n",
      "                   'the range of ratios yielded by various dose/duration pairs '\n",
      "                   '(Figure S6). As seen in Figure c, there exists an inverse '\n",
      "                   'relationship between dose and duration, with increases in '\n",
      "                   'either variable causing a larger fraction of the '\n",
      "                   'population to transition to the high GFP state. Therefore, '\n",
      "                   'similar fractional responses can be obtained through '\n",
      "                   'multiple induction routes. For example, as demonstrated in '\n",
      "                   'the middle two panels of Figure c, 8 h of 25 ng/mL '\n",
      "                   'induction and 4 h of 30 ng/mL both produced about 59% of '\n",
      "                   'ON cells. Therefore, this pulsed induction method is '\n",
      "                   'experimentally verified to be able to produce tightly '\n",
      "                   'controlled yeast population ratios.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'The model fit of the experimental data (Figure a) provides '\n",
      "                   'insight into the range of doses and durations appropriate '\n",
      "                   'for producing specific ratios. For example, with a 20 '\n",
      "                   'ng/mL ATc induction, it took between 18 and 24 h pulses to '\n",
      "                   'have greater than 70% of the cells to transition to the '\n",
      "                   'high GFP state, whereas a 40 ng/mL induction produced the '\n",
      "                   'same transition with a 4 h pulse. Generally, we observe '\n",
      "                   'that larger doses produce fast switching dynamics, and '\n",
      "                   'smaller doses required progressively longer durations to '\n",
      "                   'produce similar switch percentages. As evidenced by the '\n",
      "                   'very long durations for the 20 ng/mL dose, induction '\n",
      "                   'pulses near the bifurcation point could require durations '\n",
      "                   'of a day or more to cause a majority of cells to '\n",
      "                   'transition to high GFP, while ATc doses below 15 ng/mL '\n",
      "                   'have essentially no effect within the experimental time '\n",
      "                   'frame. With the use of experimentally validated model and '\n",
      "                   'parameters, more simulation data were analyzed to '\n",
      "                   'determine the robustness of the system to temporal '\n",
      "                   'perturbation of the pulse length. For this, we looked at '\n",
      "                   'the difference in time required to cause 30% and 70% of '\n",
      "                   'the population to transition, which we termed tuning range '\n",
      "                   '(TR = T 70 – T 30). Figure b shows that while lower doses '\n",
      "                   'require longer durations to achieve a desired ratio '\n",
      "                   '(higher vertical positions of colored bars toward the '\n",
      "                   'left), these doses also have a bigger TR (longer vertical '\n",
      "                   'span of colored bars), suggesting that the ratio tuning in '\n",
      "                   'this region is more robust against temporal variation of '\n",
      "                   'induction pulses. For example, a 25 ng/mL dose generates '\n",
      "                   '30% high GFP with a 4 h pulse and 70% with a 7 h pulse, '\n",
      "                   'resulting in a TR of 3 h. For 20 ng/mL induction, T 30 and '\n",
      "                   'T 70 both increase to 8 and 13 h, respectively, resulting '\n",
      "                   'in an overall larger TR of 5 h, indicating that there is '\n",
      "                   'more room for pulse-length error if a specific ratio '\n",
      "                   'within this range is desired. Conversely, larger doses, '\n",
      "                   'due to their fast switching dynamics, leave little room '\n",
      "                   'for error if a specific ratio is desired, with T 30 and T '\n",
      "                   '70 being nearly the same. Results for on–off transitions '\n",
      "                   'show a similar relationship between induction dose and '\n",
      "                   'required duration (Figure S7), and E. coli exhibits '\n",
      "                   'similar, though noisier, switching behavior as well '\n",
      "                   '(Figure S8).'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Mathematical modeling reveals the relationship between '\n",
      "                   'dose and duration needed for precise ratio control. (a) '\n",
      "                   'Experimental data (circles) were used to fit the '\n",
      "                   'stochastic model (lines) for multiple dose-duration pairs, '\n",
      "                   'showing strong agreement between the model and '\n",
      "                   'experimental results. Experimental data are the mean of '\n",
      "                   'four or five replicates’ medians ± SD and the simulation '\n",
      "                   'data are the mean of 1000 simulated cells. (b) Further '\n",
      "                   'simulations show induction duration needed for a specific '\n",
      "                   'ratio increases exponentially as dose decreases until near '\n",
      "                   'the bifurcation point at roughly 20 ng/mL. Additionally, '\n",
      "                   'the tuning range from when a small portion of the '\n",
      "                   'population has switched states (teal) to when a large '\n",
      "                   'fraction has transitioned (red) is large at low doses but '\n",
      "                   'shrinks with increasing doses.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'To develop a complete understanding of the relationship '\n",
      "                   'between noise and ratio control strategies across the '\n",
      "                   'noise spectrum, we employed an in silico approach that '\n",
      "                   'allows us to adjust the noisiness of gene expression while '\n",
      "                   'holding other system parameters constant. The model has '\n",
      "                   'been shown capable of recapitulating experimental results '\n",
      "                   'accurately under various conditions. It therefore serves '\n",
      "                   'as an appropriate tool to conduct thorough in silico '\n",
      "                   'explorations to meaningfully ratio control outcomes '\n",
      "                   'between high and low noise scenarios.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'To investigate the effect of noise on system '\n",
      "                   'responsiveness to induction, simulations, like those shown '\n",
      "                   'in Figure S9, were carried out in low and high noise '\n",
      "                   'settings. Figure a shows that increasing noise reduces the '\n",
      "                   'time required for a population to transition to a new '\n",
      "                   'steady state ratio. The orange line indicates the steady '\n",
      "                   'state percentage of the population which will transition '\n",
      "                   'to the high-GFP state at a given ATc concentration. All '\n",
      "                   'distributions between 0 and 100% are represented within '\n",
      "                   'the circuit’s bistable region, with the ATc concentrations '\n",
      "                   'resulting in 30% and 70% on cells (gray region) defining a '\n",
      "                   'region of broad tunability. The steady state is strictly 0 '\n",
      "                   'or 100% to the left or right of the bistable region, '\n",
      "                   'respectively. Because the time required to reach a steady '\n",
      "                   'state approaches infinity in the absence of noise, we '\n",
      "                   'measured the time needed for the population to reach half '\n",
      "                   'of the steady state ON percentage (T halfmax). In the low '\n",
      "                   'noise setting (light blue line), T halfmax is very long '\n",
      "                   'within the bistable region, with times of 50 h or more for '\n",
      "                   'steady state ratios below 80%, suggesting it is almost '\n",
      "                   'experimentally impossible. This is consistent with our '\n",
      "                   'previous studies in yeast. In the high noise setting (dark '\n",
      "                   'blue line), T halfmax is universally reduced. Beyond the '\n",
      "                   'bistable region, noise plays a less prominent role, with '\n",
      "                   'the high and low noise conditions resulting in almost '\n",
      "                   'identical transition times. These results indicate that '\n",
      "                   'constant induction within the bistable region is only a '\n",
      "                   'suitable ratio control strategy for systems with high '\n",
      "                   'enough noise to induce frequent and spontaneous state '\n",
      "                   'transitions, which is the case for E. coli.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Gene expression noise determines appropriate method for '\n",
      "                   'achieving predictable ratio control. (a) High noise allows '\n",
      "                   'ratio control with constant induction. The orange line '\n",
      "                   'indicates the steady state population level average '\n",
      "                   'expression of the system, with the gray region bounded by '\n",
      "                   'constant induction dosage needed for 30% and 70% high GFP. '\n",
      "                   'Blue lines indicate T halfmax for low noise (light blue) '\n",
      "                   'and high noise (dark blue) environments. High gene '\n",
      "                   'expression noise reduces the time required for the '\n",
      "                   'population to transition from off to the steady state '\n",
      "                   'determined by inducer concentration. The hysteresis curve '\n",
      "                   '(black) is included for visual reference. (b) Low noise '\n",
      "                   'requires transient induction. Network transition speed '\n",
      "                   '(SR, red, left y-axis) and robustness (TR, dashed purple, '\n",
      "                   'right y-axis) of ratio control for a range of induction '\n",
      "                   'dosages. These curves divide the induction space into five '\n",
      "                   'regions. Regions A, C, and E are unsuitable for generating '\n",
      "                   'controlled ratios because cells do not transition (A), '\n",
      "                   'transition too slowly (C), or lack robust ratio control '\n",
      "                   '(E). Precise ratios can be attained in region B if noise '\n",
      "                   'is high enough to increase SR to an acceptable level. '\n",
      "                   'Region D is suitable for ratio control if both SR and TR '\n",
      "                   'are large enough, determined empirically. Here, region D '\n",
      "                   'meets the criteria SR > 0.25 and TR > 5 h in the low noise '\n",
      "                   'setting. Ω = 2 for high noise and Ω = 10 for low noise.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Because of the long transition times in the low noise '\n",
      "                   'setting, the ratio control strategy for these cells '\n",
      "                   'becomes transient induction outside the bistable region. '\n",
      "                   'To compare the impacts of noise from multiple '\n",
      "                   'perspectives, we developed criteria to measure system '\n",
      "                   'responsiveness. Stimulus responsivity (SR = 1/T 30) is a '\n",
      "                   'measure of the speed of transitions in response to a '\n",
      "                   'stimulus. A higher SR value means faster ratio control. '\n",
      "                   'Along with TR, both metrics can be calculated for pulsed '\n",
      "                   'inductions as used in Figure.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Both TR and SR were computed for simulations of pulsed '\n",
      "                   'inductions for various inducer concentrations in low noise '\n",
      "                   '(Figure b) and high noise (Figure S10) settings. '\n",
      "                   'Generally, noise causes a mild universal increase in SR, '\n",
      "                   'but differences in TR are only notable within the bistable '\n",
      "                   'region. As can be seen in Figure b, while the system’s '\n",
      "                   'responsiveness (SR, red line) increases asymptotically as '\n",
      "                   'the dose increases, tolerance to error (TR, purple line) '\n",
      "                   'decreases. These data roughly divide the system’s '\n",
      "                   'induction range into five broad categories, labeled A–E. '\n",
      "                   'Regions A, C, and E are not suitable for either method of '\n",
      "                   'ratio control, but regions B and D (shaded gray) may be '\n",
      "                   'appropriate under certain conditions. Region A is '\n",
      "                   'unsuitable for ratio control because there is neither '\n",
      "                   'enough noise nor induction strength to cause a substantial '\n",
      "                   'fraction of the population to turn on. The boundaries of '\n",
      "                   'region B are fixed mathematically by the system’s '\n",
      "                   'parameters and indicate the range of constant induction '\n",
      "                   'that could result in a broad range of precisely tuned '\n",
      "                   'ratios. However, as discussed, large enough intracellular '\n",
      "                   'noise is required to attain these ratios within a '\n",
      "                   'reasonably short time frame. Regions C–E cover induction '\n",
      "                   'levels only appropriate for pulsed induction, because '\n",
      "                   'constant induction at these levels will eventually cause '\n",
      "                   'most cells to turn on. The borders separating these three '\n",
      "                   'regions are determined by SR and TR: the C–D border set by '\n",
      "                   'SR, and D–E border set by TR. The location of these '\n",
      "                   'borders should vary between organisms and could be '\n",
      "                   'determined by the cell’s physiological constraints and '\n",
      "                   'needs in their natural settings. Here, the constraints of '\n",
      "                   'SR > 0.25 and TR > 5 h provide a good inducer range D that '\n",
      "                   'balances speed and accuracy for the low noise condition. '\n",
      "                   'In region C, the system is highly tunable (large TR) but '\n",
      "                   'responds too slowly (small SR). In region E, the response '\n",
      "                   'time is fast (large SR), but tuning accuracy is lost '\n",
      "                   '(small TR). Region D finds a compromise between SR and TR '\n",
      "                   'constraints and is therefore appropriate for pulsed '\n",
      "                   'induction ratio control. The borders of region B are fixed '\n",
      "                   'mathematically and are suitable for constant induction '\n",
      "                   'ratio control. Those of region D are softer and produce '\n",
      "                   'tightly controlled ratios via pulsed induction. These '\n",
      "                   'borders are limited more by experimental constraints, and '\n",
      "                   'those shown in Figure demonstrate the inherent trade-off '\n",
      "                   'between speed and accuracy.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'In this work, we have used the mutual inhibition toggle '\n",
      "                   'switch–a synthetic version of a common genetic memory '\n",
      "                   'motif–to explore the temporal aspects of cell fate '\n",
      "                   'determination. Extracellular factors can drive a fraction '\n",
      "                   'of a population to switch phenotype and these factors '\n",
      "                   'interplay with intracellular gene regulation networks and '\n",
      "                   'noise to determine population response. In the case of E. '\n",
      "                   'coli, with noisy gene expression dynamics, applying '\n",
      "                   'constant induction within the bistable region was enough '\n",
      "                   'to elicit state-switching. With the less noisy kinetics '\n",
      "                   'observed in S. cerevisiae, on the other hand, cells '\n",
      "                   'retained their steady state behavior unless temporarily '\n",
      "                   'forced out of the bistable region with a pulse of '\n",
      "                   'induction. Through both methods, it was demonstrated that '\n",
      "                   'control of population ratios could be achieved with a high '\n",
      "                   'degree of precision.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'In the case of transiently introduced pulses of '\n",
      "                   'extracellular factors, we show that there exists an '\n",
      "                   'inverse relationship between the strength of the stimulus '\n",
      "                   'and duration for which it is needed. A target ratio can be '\n",
      "                   'achieved reliably either with a strong pulse for a short '\n",
      "                   'duration or vice versa. Under this framework, the only '\n",
      "                   'limitation for achieving precise ratio control is the '\n",
      "                   'temporal resolution imposed by the physical constraints in '\n",
      "                   'removing cells from the forcing stimulus. We also observed '\n",
      "                   'a trade-off between system response speed and tuning '\n",
      "                   'robustness, quantified in the concepts of SR and TR. Lower '\n",
      "                   'pulse doses tended to be more robust to temporal variation '\n",
      "                   'in pulse length but required long durations to reach a '\n",
      "                   'desired ratio, while the opposite was true for high doses. '\n",
      "                   'This trade-off between speed and accuracy is found '\n",
      "                   'throughout intrinsically stochastic biological contexts. '\n",
      "                   'Furthermore, gene expression noise acts as a global '\n",
      "                   'regulator of state switching. Increasing noise reduces '\n",
      "                   'transition times for both constant and pulsed inductions, '\n",
      "                   'increasing speed but resulting in a small reduction in '\n",
      "                   'tuning accuracy. Through this mechanism, noise levels '\n",
      "                   'determine whether constant induction will cause '\n",
      "                   'transitions within a reasonable time and determine the '\n",
      "                   'final dose at which pulsed induction should be set. '\n",
      "                   'Intracellular noise, therefore, is integral for the choice '\n",
      "                   'of ratio control strategy as well as for the chosen '\n",
      "                   'strategy’s implementation.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'In addition to the specific network studied, we also '\n",
      "                   'introduce a framework to analyze complex cellular '\n",
      "                   'behaviors involving a temporal component. Timing of '\n",
      "                   'cellular processes is becoming an increasingly important '\n",
      "                   'area of study, and analogies can be drawn to applied '\n",
      "                   'fields of study. Though significantly more complex, the '\n",
      "                   'methods for deriving specific phenotypes from pluripotent '\n",
      "                   'progenitors are similar to the method we employ here. '\n",
      "                   'Pluripotent cells are grown in a cocktail of growth '\n",
      "                   'factors for specific periods of time, sometimes '\n",
      "                   'sequentially, to force differentiation down a desired '\n",
      "                   'path. While this work has been biologically and '\n",
      "                   'empirically driven in the past, we suggest that a '\n",
      "                   'mathematical approach may yield further insight into '\n",
      "                   'directed differentiation methods.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Additionally, we compared ratio tuning strategies between '\n",
      "                   'organisms, showing that response times and noise profiles '\n",
      "                   'differed vastly between E. coli and S. cerevisiae. Because '\n",
      "                   'these model organisms are often used to draw inferences '\n",
      "                   'about gene networks more generally, understanding the '\n",
      "                   'nuances of their behavior and how they differ from other '\n",
      "                   'cell types, like mammalian systems, is important. '\n",
      "                   'Adjusting noise profiles in silico is a useful tool, but '\n",
      "                   'it is no substitute for studying the phenomenon in living '\n",
      "                   'systems. By better understanding how noise is manifest in '\n",
      "                   'natural circuits, we will be better able to devise '\n",
      "                   'strategies to utilize and control noise to make more '\n",
      "                   'reliable gene networks.'},\n",
      "          {'section': ['result', 'discussion'],\n",
      "           'text': 'Through this work we have shown the importance of '\n",
      "                   'considering the temporal evolution and expression noise of '\n",
      "                   'a system when analyzing its differentiation dynamics. We '\n",
      "                   'developed and leveraged a mathematical understanding of '\n",
      "                   'the bistable toggle switch to achieve robust control of '\n",
      "                   'fractional differentiation ratios. Further work along '\n",
      "                   'these lines could have wide-ranging applications in '\n",
      "                   'countering bacterial persistence, developmental or stem '\n",
      "                   'cell biology, therapeutics, and provide guidance for de '\n",
      "                   'novo gene network synthesis.'},\n",
      "          {'section': ['method'], 'text': 'Methods'},\n",
      "          {'section': ['method'], 'text': 'Plasmids and Cell Strains'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'E. coli experiments were performed with K12 MG1655 '\n",
      "                   '(American Type Culture Collection, ATCC, #700926) modified '\n",
      "                   'with (ΔLacI ΔAraC) deletions. The toggle switch plasmid, '\n",
      "                   'pKDL, was provided generously as a gift from James '\n",
      "                   'Collins. All yeast experiments were performed in YPH500 '\n",
      "                   'cells (Stratagene). The genomically integrated toggle '\n",
      "                   'system was developed previously by our group in '\n",
      "                   'collaboration with James Collins.'},\n",
      "          {'section': ['method'], 'text': 'Flow Cytometry and Data Analysis'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'All cell measurements were taken with a Becton Dickinson '\n",
      "                   '(BD) Accuri C6 flow cytometer. Front scatter (FSC-A) and '\n",
      "                   'side scatter (SSC-A) were used to gate cellular '\n",
      "                   'populations. Only a very course gating was used which '\n",
      "                   'removed debris smaller than the cell size but maintained '\n",
      "                   'the full range of population size variation. Samples were '\n",
      "                   'run on high flow rate to 10 000 captured events. The FL-1 '\n",
      "                   'channel (488 nm excitation; 530 ± 15 nm filter) was used '\n",
      "                   'to measure GFP fluorescence. Data were analyzed using '\n",
      "                   'MATLAB (Mathworks, Inc.) run on a personal computer.'},\n",
      "          {'section': ['method'], 'text': 'E. coli Experiments'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'The pKDL toggle switch was transformed into the E. coli '\n",
      "                   'using a transformation kit (Zymo Research) and selected '\n",
      "                   'for by plating on Luria Broth (LB) agar (Sigma-Aldrich) '\n",
      "                   'plates with added kanamycin (Sigma-Aldrich). Cells were '\n",
      "                   'picked the day prior to performing experiments and '\n",
      "                   'cultured in 5 mL of LB medium (Sigma-Aldrich) with '\n",
      "                   'kanamycin (Sigma-Aldrich). The following day, cell density '\n",
      "                   'was measured with flow cytometry and diluted to 1.5 '\n",
      "                   'cells/μL in fresh LB media with kanamycin.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'For hysteresis experiments, after 2 h, cells were '\n",
      "                   'rediluted into medium with anhydrotetracycline (ATc; '\n",
      "                   'Sigma-Aldrich) forcing them to the initial off (ATc = 0; '\n",
      "                   'the cells favored the off state after overnight growth) or '\n",
      "                   'initial on (ATc = 20 ng/mL) states. These were maintained '\n",
      "                   'with hourly dilutions for 3 h, then rediluted in medium '\n",
      "                   'containing variable ATc levels (0, 0.5, 1, 1.5, 2, 2.5, 3, '\n",
      "                   '3.5, and 4 ng/mL). When diluting from a high ATc '\n",
      "                   'concentration to a lower one, media volumes of different '\n",
      "                   'concentrations were mixed to avoid the potential shock of '\n",
      "                   'centrifugation and washing. For example, to go from 20 '\n",
      "                   'ng/mL to 4 ng/mL ATc, 1 part of the original culture was '\n",
      "                   'added to 4 parts at 0 ng/mL, yielding a final '\n",
      "                   'concentration of 4 ng/mL. To dilute from initial on to 0 '\n",
      "                   'ng/mL, cells were centrifuged and washed with fresh LB '\n",
      "                   'before redilution. The cultures were rediluted in fresh '\n",
      "                   'medium to 1.5 cell/μL every 3 h to avoid overgrowth and '\n",
      "                   'monitor healthy growth. They were measured every 3 h until '\n",
      "                   'they had achieved steady state expression at 15 h.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'For dose/duration experiments (Figure S8), cells were '\n",
      "                   'again forced to their initial state for 3 h, then '\n",
      "                   'rediluted with medium containing the desired dose of ATc. '\n",
      "                   'At each time point, a portion of the culture was run on '\n",
      "                   'flow cytometry and another portion was rediluted to within '\n",
      "                   'the toggle’s bistable range (0.125 ng/mL) using a similar '\n",
      "                   'fractional volume method of increasing or decreasing the '\n",
      "                   'concentration. The longest pulse given was 4 h, and all '\n",
      "                   'cultures were maintained with hourly dilutions. All '\n",
      "                   'cultures were tested via flow cytometry 4 h after the '\n",
      "                   'beginning of the pulse and again 1 h later to ensure that '\n",
      "                   'the population had reached steady state expression.'},\n",
      "          {'section': ['method'], 'text': 'S. cerevisiae Experiments'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'A single copy of the toggle switch was integrated into the '\n",
      "                   'yeast genome as described in a previous work. Confirmed '\n",
      "                   'clones were streaked onto 2% glucose YPD agar plates '\n",
      "                   '(Sigma-Aldrich). Colonies were picked from these plates 42 '\n",
      "                   'h prior to the start of the experiment and grown in 5 mL '\n",
      "                   'of YPD medium. After 8 h, cultures were monitored by flow '\n",
      "                   'cytometry and rediluted to 1500 cells/mL in fresh YPD and '\n",
      "                   'allowed to grow overnight. After 12 h, cells were measured '\n",
      "                   'again and rediluted to 5000 cells/mL into yeast medium '\n",
      "                   'with 2% galactose (Sigma-Aldrich) and 1% raffinose '\n",
      "                   '(Sigma-Aldrich) with appropriate ATc to induce the initial '\n",
      "                   'off (0 ng/mL ATc) or initial on (50 ng/mL) states. These '\n",
      "                   'were measured and rediluted again 12 h later and allowed '\n",
      "                   'to grow overnight before the experiment was begun. For '\n",
      "                   'hysteresis experiments, initial off and initial on cells '\n",
      "                   'were diluted into varying concentrations of ATc (0, 0.5, '\n",
      "                   '1, 2, 4, 8, 14, 20, 30, 40, 50 ng/mL). Cells were diluted '\n",
      "                   'to 5000 cells/mL every 12 h and measured via flow '\n",
      "                   'cytometry at 24 and 48 h. For dose/duration experiments, '\n",
      "                   'cultures were induced with ATc and at the end of each '\n",
      "                   'duration a portion of the culture was measured with flow '\n",
      "                   'cytometry and a portion was diluted down to an ATc '\n",
      "                   'concentration within the toggle’s bistable range (8 '\n",
      "                   'ng/mL). Cells were measured and rediluted every 12 h after '\n",
      "                   'the start of the initial dose, with the final measure '\n",
      "                   'being at least 12 h after the end of the initial dose, to '\n",
      "                   'ensure that cells had reached steady state expression.'},\n",
      "          {'section': ['method'], 'text': 'Modeling'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'We used the model proposed in previous work. The ODEs are '\n",
      "                   '1 [ L ] ′ = τ { c r l + 1.0 1.0 + [ ([ T ] k t) (1.0 + A T '\n",
      "                   'c k a t c k t [ T ]) − m ] n t (c i l − c r l) − d e l t a '\n",
      "                   '[ L ] } 2 [ T ] ′ = τ { c r t + 1.0 1.0 + ([ L ] k l) n l '\n",
      "                   '· (c i t − c r t) − d e l t a [ T ] } Where [L] and [T] '\n",
      "                   'are the concentration of LacI and TetR. LacI is '\n",
      "                   'coexpressed with GFP, and thus it was used '\n",
      "                   'interchangeably. crl and cil are the production rate of '\n",
      "                   'LacI when the promoter is repressed or induced, '\n",
      "                   'respectively, while crt and cit are the production rate of '\n",
      "                   'TetR when the promoter is repressed or induced, '\n",
      "                   'respectively. k t represents the active TetR concentration '\n",
      "                   'needed to make this probability 50%, and nt describes the '\n",
      "                   'nonlinearity of this inhibition. k l represents the LacI '\n",
      "                   'concentration needed to make the promoter bound by LacI '\n",
      "                   '50% of the time, and nl describes the nonlinearity of this '\n",
      "                   'inhibition. m is the Hill coefficient of the Hill '\n",
      "                   'function, which is used to describe the relationship '\n",
      "                   'between the active ratio of repressor TetR and the ATc '\n",
      "                   'inducer concentration. Here τ is the time scale of the '\n",
      "                   'system. The detail of the model construction can be found '\n",
      "                   'in our previous work.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'For the stochastic simulation, the concentration of each '\n",
      "                   'molecular species is converted to its number; that is, x = '\n",
      "                   '[ x ]\\xadΩ, where Ω is a system size factor. Table S1 '\n",
      "                   'lists all the reactions involved. The τ-leap-based '\n",
      "                   'stochastic Gillespie algorithm is used for the stochastic '\n",
      "                   'simulation. The noise level is set by Ω, which is set to 3 '\n",
      "                   'for E. coli system and 10 for S. cerevisiae. The '\n",
      "                   'quasi-potential landscape in Figure d and Figure b is '\n",
      "                   'defined as U (x) = −log\\xad(P ss (X)) where P ss (X) is '\n",
      "                   'the probability distribution from stochastic simulation '\n",
      "                   'with 1000 cells for 500 model hours. Figure d and Figure b '\n",
      "                   'show the resultant quasipotential landscape corresponding '\n",
      "                   'to a one-dimensional projection on the LacI axis. The '\n",
      "                   'bifurcation diagrams are generated with Oscill8 '\n",
      "                   '(http://oscill8.sourceforge.net/).'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'We searched the parameter space with a customized '\n",
      "                   'Metropolis algorithm (see the following section for '\n",
      "                   'details) to fit various experimental data, including the '\n",
      "                   'hysteresis curves (Figures b and a) and the fractions of '\n",
      "                   'the population to transition to the high GFP state '\n",
      "                   '(Figures e and a). It is noted that not all the parameters '\n",
      "                   'for E. coli and yeast are the same since we found '\n",
      "                   'different hysteresis curves in the two systems. It is an '\n",
      "                   'irreversible bistable switch in E. coli (Figure b), while '\n",
      "                   'it is a reversible bistable switch in S. cerevisiae '\n",
      "                   '(Figure a). To preserve the bistability for both systems '\n",
      "                   'and show the difference between them, we kept most of the '\n",
      "                   'parameters the same for the two systems as the synthetic '\n",
      "                   'circuit is topologically the same. We estimated the '\n",
      "                   'parameters based on our previous work to make both systems '\n",
      "                   'bistable and then used the Metropolis algorithm to search '\n",
      "                   'the parameters that control the bistable ranges (including '\n",
      "                   'k act, k t, and k l) and the time scale (τ) separately to '\n",
      "                   'fit various experimental data. The estimated and fitted '\n",
      "                   'parameters can be found in Table S2. In Figure, we used '\n",
      "                   'the parameters for S. cerevisiae under different noise '\n",
      "                   'levels to study the general strategies to achieve a '\n",
      "                   'predictable ratio.'},\n",
      "          {'section': ['method'], 'text': 'Parameter Searching'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'The procedure for parameter searching with a customized '\n",
      "                   'Metropolis algorithm was as follows. 1. Choose a parameter '\n",
      "                   'set Para 0 = [ k act, k t, k l, τ] and calculate its '\n",
      "                   'fitting score, Score 0. a. For the E. coli system, for '\n",
      "                   'each ATc dose in the Figure b and e, the mean value of '\n",
      "                   '1000 simulated cells will be calculated. Then the '\n",
      "                   'summation of the absolute values of the difference between '\n",
      "                   'the simulation data and experiments, i.e., mean absolute '\n",
      "                   'error (MAE), is calculated as the fitting score b. For the '\n",
      "                   'yeast system, for each ATc dose in Figure a and each ATc '\n",
      "                   'dose/duration combination point in Figure a, the mean '\n",
      "                   'value of 1000 simulated cells will be calculated. Then the '\n",
      "                   'summation of the absolute values of the difference between '\n",
      "                   'the simulation data and experiments, i.e., mean absolute '\n",
      "                   'error (MAE), is calculated as the fitting score. That is, '\n",
      "                   'score = ∑ | Y s i m − Y exp | n, where Y sim is the '\n",
      "                   'simulation results results and Y exp is the experimental '\n",
      "                   'data. 2. Generate a new parameter set Para 1 = Para 0 (1 – '\n",
      "                   'Δ P + ε2Δ P), where Δ P specifies the maximum percentage '\n",
      "                   'of change per step and ε is a vector of uniformly '\n",
      "                   'distributed random number in the interval (0,1). 3. '\n",
      "                   'Calculate the fitting score Score 1 with the parameter set '\n",
      "                   'Para 1. 4. Calculate the acceptance probability p = e − '\n",
      "                   '(Score 1 − Score 0)/T. Generate a random number γ 1 from '\n",
      "                   'with uniform distribution between 0 and 1. Update Para 0 = '\n",
      "                   'Para 1 if γ 1 < p. Otherwise, reject the step k to k + 1. '\n",
      "                   '5. Update the best score Score best = min\\xad(Score '\n",
      "                   '1,Score 0) and the corresponding best parameter set Para '\n",
      "                   'best 6. Update step number k. If k is larger than a '\n",
      "                   'maximum step number, stop. Otherwise return to step 1.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Choose a parameter set Para 0 = [ k act, k t, k l, τ] and '\n",
      "                   'calculate its fitting score, Score 0.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'For the E. coli system, for each ATc dose in the Figure b '\n",
      "                   'and e, the mean value of 1000 simulated cells will be '\n",
      "                   'calculated. Then the summation of the absolute values of '\n",
      "                   'the difference between the simulation data and '\n",
      "                   'experiments, i.e., mean absolute error (MAE), is '\n",
      "                   'calculated as the fitting score'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'For the yeast system, for each ATc dose in Figure a and '\n",
      "                   'each ATc dose/duration combination point in Figure a, the '\n",
      "                   'mean value of 1000 simulated cells will be calculated. '\n",
      "                   'Then the summation of the absolute values of the '\n",
      "                   'difference between the simulation data and experiments, '\n",
      "                   'i.e., mean absolute error (MAE), is calculated as the '\n",
      "                   'fitting score. That is, score =, where Y sim is the '\n",
      "                   'simulation results results and Y exp is the experimental '\n",
      "                   'data.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Generate a new parameter set Para 1 = Para 0 (1 – Δ P + '\n",
      "                   'ε2Δ P), where Δ P specifies the maximum percentage of '\n",
      "                   'change per step and ε is a vector of uniformly distributed '\n",
      "                   'random number in the interval (0,1).'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Calculate the fitting score Score 1 with the parameter set '\n",
      "                   'Para 1.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Calculate the acceptance probability. Generate a random '\n",
      "                   'number γ 1 from with uniform distribution between 0 and 1. '\n",
      "                   'Update Para 0 = Para 1 if γ 1 < p. Otherwise, reject the '\n",
      "                   'step k to k + 1.'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Update the best score Score best = min\\xad(Score 1,Score '\n",
      "                   '0) and the corresponding best parameter set Para best'},\n",
      "          {'section': ['method'],\n",
      "           'text': 'Update step number k. If k is larger than a maximum step '\n",
      "                   'number, stop. Otherwise return to step 1.'}],\n",
      " 'electron_pub_date': '05/14/2019',\n",
      " 'history': [{'event': 'received', 'time': '01/24/2019'},\n",
      "             {'event': 'asap', 'time': '05/29/2019'},\n",
      "             {'event': 'just-accepted', 'time': '05/14/2019'}],\n",
      " 'internal_id': 'a984433005',\n",
      " 'is_research': True,\n",
      " 'issue_pub_date': '06/21/2019',\n",
      " 'keywords': ['synthetic biology',\n",
      "              'ratio tuning',\n",
      "              'stochasticity',\n",
      "              'fractional differentiation',\n",
      "              'cell fate'],\n",
      " 'type': 'Research Article'}\n"
     ]
    }
   ],
   "source": [
    "# print one sample for inspection\n",
    "pprint(processed_files[\"sb9b00030\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Correction': 4,\n",
      " 'Editorial': 27,\n",
      " 'In This Issue': 69,\n",
      " 'Introducing Our Authors': 66,\n",
      " 'Letter': 368,\n",
      " 'Research Article': 907,\n",
      " 'Review': 27,\n",
      " 'Technical Note': 82,\n",
      " 'Tutorial': 4,\n",
      " 'Viewpoint': 43}\n",
      "\n",
      "{'accepted': 1,\n",
      " 'asap': 1398,\n",
      " 'issue-pub': 483,\n",
      " 'just-accepted': 1285,\n",
      " 'received': 1597}\n"
     ]
    }
   ],
   "source": [
    "# show some stats\n",
    "type_count = {}\n",
    "event_count = {}\n",
    "for k, article_info in processed_files.items():\n",
    "    article_type = article_info[\"type\"]\n",
    "    if not article_type in type_count:\n",
    "        type_count[article_type] = 0\n",
    "    type_count[article_type] += 1\n",
    "    \n",
    "    events = article_info[\"history\"]\n",
    "    for e in events:\n",
    "        if not e[\"event\"] in event_count:\n",
    "            event_count[e[\"event\"]] = 0\n",
    "        event_count[e[\"event\"]] += 1\n",
    "pprint(type_count)\n",
    "print()\n",
    "pprint(event_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Genetic Sequences Extraction\n",
    "\n",
    "For this part we want to go through all the supplementary files in the data dump and recover genetic sequences from them. This requires handling of different file type and using the sbol api. Once we extracted the genetic sequences, we will store the sequence data in fasta format in `processed_files[pub_num][suppl_files][sequences]` for later use."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2.1: Unzipping\n",
    "\n",
    "Some supplementary files are compressed into zip files, we want to unzip them first before doing anything. Because we only need to do it once, the code below is commented out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find out all the zip files and unzip them\n",
    "# do this once when there is new data\n",
    "# suppl_files_zip = []\n",
    "# collect_files(input_path, suppl_files_zip, pattern=\".*\\.zip$\", collect_dirs=False, min_depth=3)\n",
    "# print(\"zip files: %d\" % len(suppl_files_zip))\n",
    "# for zip_file in suppl_files_zip:\n",
    "#     zip_file_dir = re.sub(\"/[^/]*$\", \"\", zip_file)\n",
    "#     res = sp.run([\"unzip\", \"-n\", zip_file, \"-d\", zip_file_dir])\n",
    "#     if res.returncode != 0:\n",
    "#         print(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2.2: Finding Paths to All Supplementary Files\n",
    "\n",
    "Unlike the locations of article xml files, the structures of suppl folder is unpreditable. Therefore we need to recursively find all the files in the suppl folder. We store the absolute path to all supplementary files in a list called `suppl_files_all`. After we have the list, we need to put this information into the `processed_file` so that we can achive two goals: 1) to query supplementary files using publication number, and 2) to locate supplementary files relative to `input_path` using publication number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all suppl files\n",
    "suppl_files_all = []\n",
    "collect_files(input_path, suppl_files_all, pattern=\"\", collect_dirs=False, min_depth=3)\n",
    "suppl_files_all = [x for x in suppl_files_all if not re.match(\".*__MACOSX.*\", x)]\n",
    "print(\"suppl files: %d\" % len(suppl_files_all))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attach all the suppl path to processed files\n",
    "# clear the suppl list of each article first to make the following code idempotent\n",
    "for article_info in processed_files.values():\n",
    "    article_info[\"suppl_files\"] = []\n",
    "    \n",
    "for path in suppl_files_all:\n",
    "    path = path.split(\"/\")\n",
    "    \n",
    "    # get basic attrib\n",
    "    suppl_filename = path[-1]\n",
    "    suppl_dir = input_path.split(\"/\")[-1]\n",
    "    suppl_dir_idx = 0\n",
    "    for i, item in enumerate(path):\n",
    "        if item == suppl_dir:\n",
    "            suppl_dir_idx = i\n",
    "            break\n",
    "    else:\n",
    "        assert False # we should not be here\n",
    "    pub_num = path[suppl_dir_idx + 1]\n",
    "    rpath = os.path.join(*path[suppl_dir_idx + 1:]) # relative path\n",
    "    \n",
    "    # create info dict\n",
    "    suppl_info = {\n",
    "        \"suppl_filename\": suppl_filename,\n",
    "        \"rpath\": rpath,\n",
    "        \"sequences\": None\n",
    "    }\n",
    "    \n",
    "    # push it into the processed files dict\n",
    "    if pub_num in processed_files:\n",
    "        processed_files[pub_num][\"suppl_files\"].append(suppl_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2.3 SOBL API Validation\n",
    "\n",
    "In this part, we use sbol api to find all the valid sequence files and extract the sequence into fasta format. There are some limitation on the sbol api: 1) the reponse time over network is long and 2) the api only accept file size up to 64MB. To address this problem, currently we are ignoring large files and maintaining a cache locally to increase the performance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get supplementary files by extension\n",
    "def filter_suppl_file_by_ext(all_articles, allowed_ext):\n",
    "    suppl_files_to_check = []\n",
    "    for article_info in all_articles.values():\n",
    "        for suppl_info in article_info[\"suppl_files\"]:\n",
    "            # only allow the following extension\n",
    "            ext = suppl_info[\"suppl_filename\"].split(\".\")[-1]\n",
    "            if ext in allowed_ext:\n",
    "                suppl_files_to_check.append((suppl_info, article_info))\n",
    "    return suppl_files_to_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use sbol api to validate sequence files\n",
    "def validate_sequence(file):\n",
    "    \n",
    "    # restrict file size to be less than 64mb\n",
    "    # this is an api restriction\n",
    "    file_size = os.path.getsize(file)\n",
    "    if file_size >= 64 * 2 ** 20:\n",
    "        return False, None\n",
    "    \n",
    "    # try to read the content\n",
    "    try:\n",
    "        with open(file, \"r\", encoding=\"utf-8\") as f:\n",
    "            content = f.read()\n",
    "    except UnicodeDecodeError:\n",
    "        return False, None\n",
    "    \n",
    "    # validate file\n",
    "    validator_param[\"main_file\"] = content\n",
    "    res = requests.post(sbol_validator_url, json=validator_param).json()\n",
    "    return res[\"valid\"], res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use the api to check if the file is a sequence file\n",
    "# we will cache the request result to reduce server load\n",
    "valid_res_cache = None\n",
    "if os.path.exists(valid_cache_path):\n",
    "    with open(valid_cache_path, \"rb\") as f:\n",
    "        valid_res_cache = pickle.load(f)\n",
    "else:\n",
    "    valid_res_cache = {}\n",
    "\n",
    "# create a list of all the suppl files that we need to check\n",
    "sbol_suppl_files_to_check = filter_suppl_file_by_ext(processed_files, sbol_allowed_file_type)\n",
    "\n",
    "for suppl_info, article_info in tqdm(sbol_suppl_files_to_check):\n",
    "\n",
    "    # use api or cache to get convert and store the sequence file in fasta format\n",
    "    path = os.path.join(input_path, suppl_info[\"rpath\"])\n",
    "    if not suppl_info[\"rpath\"] in valid_res_cache:\n",
    "        is_valid, data = validate_sequence(path)\n",
    "        valid_res_cache[suppl_info[\"rpath\"]] = (is_valid, data)\n",
    "    else:\n",
    "        is_valid, data = valid_res_cache[suppl_info[\"rpath\"]]\n",
    "    suppl_info[\"sequences\"] = data[\"result\"] if is_valid else None\n",
    "\n",
    "# save cache\n",
    "with open(valid_cache_path, \"wb\") as f:\n",
    "    pickle.dump(valid_res_cache, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2.4 Extract Sequences from PDF\n",
    "\n",
    "Most of the supplementary files are not valid sequence files, but a lot of them are pdf files and contain sequences. In this part we try to use the pdftotext to convert the pdf to plain text first, and then salvage sequences from the plain text.\n",
    "\n",
    "When a sequence in pdf is converted to plain text, two thing can happend: 1) the sequence is splited across multiple lines, or 2) the sequence remain in one line. The script extract the multiline sequences first and then replace them with a place holder and then extract the single line sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finds any strings that looks like a sequence in pdf\n",
    "# need to have pdftotext installed\n",
    "def extract_sequence_from_pdf(path):\n",
    "    sp.run([\"pdftotext\", \"-raw\", path, \"/tmp/sbks-pdf-tmp.txt\"]) # use tmp dir maybe it is ram so faster?\n",
    "    seqs = []\n",
    "    with open(\"/tmp/sbks-pdf-tmp.txt\", \"r\") as file:\n",
    "        lines = file.readlines()\n",
    "        text = \"\".join(lines).upper()\n",
    "        \n",
    "        # first find all the multiline sequences\n",
    "        for multi_line_seq in re.findall(\"([ATCG]{3,}(\\n[ATCG]{3,}){1,})\", text):\n",
    "            seqs.append(re.sub(\"\\n\", \"\", multi_line_seq[0]))\n",
    "            \n",
    "        # remove extracted sequences to avoid double counting\n",
    "        text = re.sub(\"([ATCG]{3,}(\\n[ATCG]{3,}){1,})\", \" ###REMOVED### \", text)\n",
    "        \n",
    "        # then find all the one line sequences\n",
    "        for single_line_seq in re.findall(\"[ATCG]{10,}\", text):\n",
    "            seqs.append(single_line_seq)\n",
    "            \n",
    "#     sp.run([\"rm\", \"tmp.txt\"])\n",
    "    return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract sequences from pdf\n",
    "# create a list of all the suppl files that we need to check\n",
    "pdf_suppl_files_to_check = filter_suppl_file_by_ext(processed_files, set([\"pdf\"]))\n",
    "\n",
    "for suppl_info, article_info in tqdm(pdf_suppl_files_to_check):\n",
    "\n",
    "    # use pdftotext to extract the sequences from pdf\n",
    "    path = os.path.join(input_path, suppl_info[\"rpath\"])\n",
    "    sequences = extract_sequence_from_pdf(path)\n",
    "    if sequences:\n",
    "        res = []\n",
    "        for i, s in enumerate(sequences):\n",
    "            res.append(f\">{article_info['internal_id']}_{suppl_info['suppl_filename']}_{i}\\n\")\n",
    "            res.append(s + \"\\n\")\n",
    "        suppl_info[\"sequences\"] = \"\".join(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Writing Results\n",
    "\n",
    "There are three things written to the disk:\n",
    "- **Processed Files:** Each entry is saved into one pickle file and named using the publication number. The pickle files are saved in `output_path`.\n",
    "- **Plain Text:** The text data of each article is saved into one txt file and named using the publication number. The txt files are saved in `txt_path`.\n",
    "- **sequences:** The sequences extracted from each article is saved into one fasta formatted txt file. They are saved in `./sequence-files` fol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the sequence files\n",
    "for pub_num, article_info in processed_files.items():\n",
    "    seq_to_write = []\n",
    "    for suppl_info in article_info[\"suppl_files\"]:\n",
    "        if not suppl_info[\"sequences\"] is None:\n",
    "            seq_to_write.append(suppl_info[\"sequences\"])\n",
    "    if len(seq_to_write) > 0:\n",
    "        with open(os.path.join(\"sequence-files\", pub_num + \"_\" + article_info[\"internal_id\"] + \".seq.txt\"), \"w\") as outfile:\n",
    "            for seq in seq_to_write:\n",
    "                outfile.write(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pickle the processed files\n",
    "for pub_num, data in tqdm(processed_files.items()):\n",
    "    with open(os.path.join(output_path, pub_num + \".pkl\"), \"wb\") as out:\n",
    "        pickle.dump(data, out)\n",
    "    text_data = [d[\"text\"] + \"\\n\" for d in data[\"body\"]]\n",
    "    \n",
    "    # save plain text\n",
    "    if data[\"is_research\"]:\n",
    "        text_data_path = os.path.join(txt_path, \"research\")\n",
    "    else:\n",
    "        text_data_path = os.path.join(txt_path, \"non-research\")\n",
    "    with open(os.path.join(text_data_path, pub_num + \".txt\"), \"w\") as out:\n",
    "        out.writelines(text_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['is_research', 'keywords', 'abstract', 'body', 'issue_pub_date', 'electron_pub_date', 'article_id', 'internal_id', 'history', 'type', 'suppl_files'])\n",
      "'Research Article'\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# check one pickle\n",
    "with open(os.path.join(output_path, \"sb9b00030.pkl\"), \"rb\") as ifile:\n",
    "    article_info = pickle.load(ifile)\n",
    "    pprint(article_info.keys())\n",
    "    pprint(article_info[\"type\"])\n",
    "    pprint(article_info[\"is_research\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
